pybind11/how-to.rst

```diff
@@ -650,105 +771,40 @@ Following :ref:`this rule <style-guide-pybind11-declare-template-wrappers>` we d
+But this only solves cross-module circular dependencies to the extent that a circular Python imports result let the second importer see the types in the partially-defined first module.
+It's not clearly that something guaranteed by the CPython interpreter (let alone the Python language), so it's much safer too just make sure any types with a circular dependency are handled by the same ``WrapperCollection`` instance and defined in the same module.
 
-    PYBIND11_MODULE(exampleTwo, mod) {
-        py::class_<ExampleBase, std::shared_ptr<ExampleBase>> clsExampleBase(mod, "ExampleBase");
-        clsExampleBase.def("someMethod", &ExampleBase::someMethod);
-
-        py::class_<ExampleTwo, std::shared_ptr<ExampleTwo>, ExampleBase> clsExampleTwo(mod, "ExampleTwo");
-        clsExampleTwo.def(py::init<>());
-        clsExampleTwo.def("someOtherMethod", &ExampleTwo::someOtherMethod);
-    }
-
-    }}  // lsst::tmpl
-
-and ``exampleThree.cc``:
-
-.. code-block:: cpp
-
-    #include "pybind11/pybind11.h"
-
-    #include "lsst/TMPL/ExampleThree.h"
-
-    namespace py = pybind11;
-    using namespace pybind11::literals;
-
-    namespace lsst {
-    namespace tmpl {
-    namespace {
-
-    template <typename T>
-    static void declareExampleThree(py::module & mod, std::string const & suffix) {
-        using Class = ExampleThree<T>;
-        using PyClass = py::class_<Class, std::shared_ptr<Class>, ExampleBase>;
-
-        PyClass cls(mod, ("ExampleThree" + suffix).c_str());
-
-        cls.def(py::init<T>());
-        cls.def("someOtherMethod", &Class::someOtherMethod);
-    }
-
-    }
-
-    PYBIND11_MODULE(exampleThree, mod) {
-        py::module::import("exampleTwo");
-
-        declareExampleThree<float>(mod, "F");
-        declareExampleThree<double>(mod, "D");
-    }
-
-    }}  // lsst::tmpl
 
 Adding pure Python members
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Sometimes it is necessary to add pure Python members to a wrapped C++ class.
-Following our :ref:`structure and naming convention <style-guide-pybind11-subpackage>` for this, we move ``exampleTwo.cc`` to a new subpackage (``exampleTwo``) and add an ``__init__.py`` file with the following content:
-
-.. code-block:: python
+Following our :ref:`structure and naming convention <style-guide-pybind11-subpackage>` for this, we'll add a new pure-Python ``_ExampleTwo.py`` module.
+Note that this name doesn't conflict with ``_ExampleTwo.cc``, because that's never compiled into a standalone Python module.
 
-    from .exampleTwo import *
-    from .exampleTwoContinued import *
-
-The pure Python code then goes into ``exampleTwo/exampleTwoContinued.py``.
-We shall use the ``continueClass`` decorator to reopen the class and add a new method:
+We'll use the ``continueClass`` decorator to reopen the class and add a new method:
 
 .. code-block:: python
 
     from lsst.utils import continueClass
 
-    from .exampleTwo import ExampleTwo
+    from ._tmpl import ExampleTwo
```

Comment:
Just checking: this should be `._tmpl` and not just `.`?

Reply:
Yes.  Both *may* work, but just `.` relies on `__init__.py` being re-imported while it is already being imported and that partially-imported state providing the `ExampleTwo` symbol.